File: front/nft_game_node_prj/wallets/solflare.ts
import * as solanaWeb3 from "@solana/web3.js";
import Wallet from "./wallet";
import bs58 from "bs58";

declare const window: Window &
  typeof globalThis & {
    solflare: any;
  };

class SolflareWallet extends Wallet {
  constructor() {
    super();
  }

  async connect(): Promise<solanaWeb3.PublicKey> {
    const resp = await window.solflare.connect();
    console.log(resp);
    return new solanaWeb3.PublicKey(window.solflare.publicKey.toString());
  }

  disconnect(): void {
    window.solflare.disconnect();
  }

  isInstalled(): Boolean {
    return window.solflare && window.solflare.isSolflare;
  }

  async signTransaction(
    tx: solanaWeb3.Transaction
  ): Promise<solanaWeb3.Transaction> {
    return await window.solflare.signTransaction(tx);
  }

  async signMessage(msg: string): Promise<String> {
    const encodedMsg = new TextEncoder().encode(msg);
    const signedMessage = await window.solflare.signMessage(encodedMsg, "utf8");
    const bs58SignedMessage = bs58.encode(signedMessage.signature);
    return bs58SignedMessage;
  }
}

export default new SolflareWallet();

File: front/lib/js/solflare.dart
import 'dart:typed_data';

import 'package:js/js.dart';
import 'package:nft_game/js/solana_web3/transaction.dart';
import 'package:nft_game/js/solana_web3/public_key.dart';

@JS('wallet.solflare.connect')
external Future<PublicKey> connect();

@JS('wallet.solflare.disconnect')
external Future<PublicKey> disconnect();

@JS('wallet.solflare.isInstalled')
external bool? isInstalled();

@JS('wallet.solflare.signTransaction')
external Future<String> signTransaction(Transaction tx);

@anonymous
@JS('wallet.solflare.SignedMessage')
class SignedMessage {
  external factory SignedMessage({Uint8List signature, PublicKey publicKey});
  external Uint8List get signature;
  external PublicKey get publicKey;
}

@JS('wallet.solflare.signMessage')
external Future<SignedMessage> signMessage(String msg);
